pkg/pool-stable-phantom/contracts/StablePoolProtocolFees.sol

```diff
@@ -24,6 +25,24 @@ import "./StableMath.sol";
 
 abstract contract StablePoolProtocolFees is StablePoolStorage, StablePoolRates, ProtocolFeeCache {
     using FixedPoint for uint256;
+    using WordCodec for bytes32;
+
+    // We store the invariant after the last join-exit, along with the amplification factor used to compute it. The
+    // amplification factor is bound by _MAX_AMP * _AMP_PRECISION, or 5e6, which fits in 23 bits. We use all remaining
+    // bits for the invariant: this is more than enough, as the invariant is proportional to the total supply, which is
+    // capped at 112 bits.
+    // The data structure is as follows:
+    //
+    // [ last join-exit amplification  | last post join-exit invariant ]
+    // [           23 bits             |            233 bits           ]
+    bytes32 private _lastJoinExitData;
+
+    uint256 private constant _LAST_POST_JOIN_EXIT_INVARIANT_OFFSET = 0;
+    uint256 private constant _LAST_POST_JOIN_EXIT_INVARIANT_SIZE = 233;
+    uint256 private constant _LAST_JOIN_EXIT_AMPLIFICATION_OFFSET = _LAST_POST_JOIN_EXIT_INVARIANT_OFFSET +
```

Comment:
I like having the offsets and sizes separated, and all hard numbers vs calculations (so that you can easily change the order, without redoing all the "equations"). But for only two things it doesn't really matter